File: integration-test/java/shared/src/main/java/com/qualityminds/lazyval/integration/shared/CouponCode.java
package com.qualityminds.lazyval.integration.shared;

import org.jspecify.annotations.Nullable;

import java.util.Objects;

public final class CouponCode {

    private final String value;

    private CouponCode(String value) {
        this.value = value;
    }

    public String value() {
        return value;
    }

    @Nullable
    public static CouponCode ofNullable(@Nullable String value) {
        if (value == null || value.isBlank()) return null;
        return new CouponCode(value.trim().toUpperCase());
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null || getClass() != obj.getClass()) return false;
        CouponCode other = (CouponCode) obj;
        return Objects.equals(value, other.value);
    }

    @Override
    public int hashCode() {
        return Objects.hash(value);
    }

    @Override
    public String toString() {
        return "CouponCode{" + value + "}";
    }
}

File: integration-test/kotlin/shared/src/main/kotlin/com/qualityminds/lazyval/integration/shared/CouponCode.kt
package com.qualityminds.lazyval.integration.shared

class CouponCode private constructor(val value: String) {

    companion object {
        @JvmStatic
        fun ofNullable(value: String?): CouponCode? {
            if (value.isNullOrBlank()) return null
            return CouponCode(value.trim().uppercase())
        }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val couponCode = other as CouponCode
        return value == couponCode.value
    }

    override fun hashCode(): Int = value.hashCode()

    override fun toString(): String = "CouponCode{$value}"
}
